Bug: `No matching configuration found` warning for code block matched only by non-universal patterns

### Environment

Node version: v20.12.0
npm version: 10.5.0
Local ESLint version: v9.3.0
Global ESLint version: no
Operating System: windows


### What parser are you using?

Default (Espree)

### What did you do?

I wanted to make a repro with `eslint-plugin-markdown`, but since [this warning](https://github.com/eslint/eslint/blob/389744be255717c507fafc158746e579ac08d77e/lib/linter/linter.js#L1986-L1993) has `line: 0` (which might be another bug), `eslint-plugin-markdown` filters it out as out-of-range. So I made a simple processor that assumes that the entire markdown file is one ` ```ts ` code block.

```js
// eslint.config.js
export default [
  {
      plugins: {
          "my-plugin": {
              processors: {
                  "my-processor": {

                      // dummy processor that assumes that the whole content of a file is a single ```ts code block 
                      preprocess(text) {
                          const lines = text.split("\n");
                          return [{
                              filename: "foo.ts",
                              text: lines.slice(1, -1).join("\n")
                          }]
                      },

                      // just increment lines to adjust for the first ```ts line
                      postprocess(messages) {
                          const retv = messages[0]; // there's always exactly 1 code block

                          retv.forEach(message => {
                              message.line++
                              message.endLine++; 
                          });

                          return retv;
                      }
                  }
              }
          }
      }
  },
  {
      files: ["**/*.md"],
      processor: "my-plugin/my-processor"
  },
  {
      files: ["**/*"],
      rules: {
          "no-undef": 2
      }
  }
];

```

`test.md`:

````md
```ts
foo
```
````


### What did you expect to happen?

Since the `.ts` virtual file doesn't have an explicit match in the config, I'd expect it to be silently ignored, so at the end I'd expect no lint errors or warnings.

### What actually happened?

```
C:\projects\tmp\tmp\test.md
  1:0  warning  No matching configuration found for C:\projects\tmp\tmp\test.md\0_foo.ts

✖ 1 problem (0 errors, 1 warning)
```

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/stackblitz-starters-hqxgng?file=eslint.config.js

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

We found this problem while discussing https://github.com/eslint/rewrite/pull/7.

I think we could just replace the following line with `return configs.getConfig(blockFilename) !== void 0;`

https://github.com/eslint/eslint/blob/389744be255717c507fafc158746e579ac08d77e/lib/eslint/eslint.js#L512

--- a/docs/src/use/configure/plugins.md
+++ b/docs/src/use/configure/plugins.md
@@ -213,7 +213,41 @@ export default [
 ];
 ```
 
-ESLint only lints named code blocks when they are JavaScript files or if they match a `files` entry in a config object. Be sure to add a config object with a matching `files` entry if you want to lint non-JavaScript named code blocks.
+ESLint only lints named code blocks when they are JavaScript files or if they match a `files` entry in a config object. Be sure to add a config object with a matching `files` entry if you want to lint non-JavaScript named code blocks. Also note that [global ignores](./ignore) apply to named code blocks as well.
+
+```js
+// eslint.config.js
+import markdown from "eslint-plugin-markdown";
+
+export default [
+
+    // applies to Markdown files
+    {
+        files: ["**/*.md"],
+        plugins: {
+            markdown
+        },
+        processor: "markdown/markdown"
+    },
+
+    // applies to all .jsx files, including jsx blocks inside of Markdown files
+    {
+        files: ["**/*.jsx"],
+        languageOptions: {
+            parserOptions: {
+                ecmaFeatures: {
+                    jsx: true
+                }
+            }
+        }
+    },
+
+    // ignore jsx blocks inside of test.md files
+    {
+        ignores: ["**/test.md/*.jsx"]
+    }
+];
+```
 
 ## Common Problems
 

--- a/lib/eslint/eslint.js
+++ b/lib/eslint/eslint.js
@@ -509,7 +509,7 @@ function verifyText({
              * @returns {boolean} `true` if the linter should adopt the code block.
              */
             filterCodeBlock(blockFilename) {
-                return configs.isExplicitMatch(blockFilename);
+                return configs.getConfig(blockFilename) !== void 0;
             }
         }
     );

--- a/tests/lib/eslint/eslint.js
+++ b/tests/lib/eslint/eslint.js
@@ -67,6 +67,12 @@ describe("ESLint", () => {
         }
     };
     const examplePreprocessorName = "eslint-plugin-processor";
+    const patternProcessor = require("../../fixtures/processors/pattern-processor");
+    const exampleMarkdownPlugin = {
+        processors: {
+            markdown: patternProcessor.defineProcessor(/```(\w+)\n(.+?)\n```(?:\n|$)/gsu)
+        }
+    };
     const originalDir = process.cwd();
     const fixtureDir = path.resolve(fs.realpathSync(os.tmpdir()), "eslint/fixtures");
 
@@ -3759,6 +3765,178 @@ describe("ESLint", () => {
                     assert.strictEqual(results[0].messages.length, 1);
                     assert.strictEqual(results[0].suppressedMessages.length, 0);
                     assert(!Object.hasOwn(results[0], "output"));
+                });
+            });
+
+            describe("matching and ignoring code blocks", () => {
+                const pluginConfig = {
+                    files: ["**/*.md"],
+                    plugins: {
+                        markdown: exampleMarkdownPlugin
+                    },
+                    processor: "markdown/markdown"
+                };
+                const text = unIndent`
+                    \`\`\`js
+                    foo_js
+                    \`\`\`
+
+                    \`\`\`ts
+                    foo_ts
+                    \`\`\`
+
+                    \`\`\`cjs
+                    foo_cjs
+                    \`\`\`
+
+                    \`\`\`mjs
+                    foo_mjs
+                    \`\`\`
+                `;
+
+                it("should by default lint only .js, .mjs, and .cjs virtual files", async () => {
+                    eslint = new ESLint({
+                        overrideConfigFile: true,
+                        overrideConfig: [
+                            pluginConfig,
+                            {
+                                rules: {
+                                    "no-undef": 2
+                                }
+                            }
+                        ]
+                    });
+                    const [result] = await eslint.lintText(text, { filePath: "foo.md" });
+
+                    assert.strictEqual(result.messages.length, 3);
+                    assert.strictEqual(result.messages[0].ruleId, "no-undef");
+                    assert.match(result.messages[0].message, /foo_js/u);
+                    assert.strictEqual(result.messages[0].line, 2);
+                    assert.strictEqual(result.messages[1].ruleId, "no-undef");
+                    assert.match(result.messages[1].message, /foo_cjs/u);
+                    assert.strictEqual(result.messages[1].line, 10);
+                    assert.strictEqual(result.messages[2].ruleId, "no-undef");
+                    assert.match(result.messages[2].message, /foo_mjs/u);
+                    assert.strictEqual(result.messages[2].line, 14);
+                });
+
+                it("should lint additional virtual files that match non-universal patterns", async () => {
+                    eslint = new ESLint({
+                        overrideConfigFile: true,
+                        overrideConfig: [
+                            pluginConfig,
+                            {
+                                rules: {
+                                    "no-undef": 2
+                                }
+                            },
+                            {
+                                files: ["**/*.ts"]
+                            }
+                        ]
+                    });
+                    const [result] = await eslint.lintText(text, { filePath: "foo.md" });
+
+                    assert.strictEqual(result.messages.length, 4);
+                    assert.strictEqual(result.messages[0].ruleId, "no-undef");
+                    assert.match(result.messages[0].message, /foo_js/u);
+                    assert.strictEqual(result.messages[0].line, 2);
+                    assert.strictEqual(result.messages[1].ruleId, "no-undef");
+                    assert.match(result.messages[1].message, /foo_ts/u);
+                    assert.strictEqual(result.messages[1].line, 6);
+                    assert.strictEqual(result.messages[2].ruleId, "no-undef");
+                    assert.match(result.messages[2].message, /foo_cjs/u);
+                    assert.strictEqual(result.messages[2].line, 10);
+                    assert.strictEqual(result.messages[3].ruleId, "no-undef");
+                    assert.match(result.messages[3].message, /foo_mjs/u);
+                    assert.strictEqual(result.messages[3].line, 14);
+                });
+
+                // https://github.com/eslint/eslint/issues/18493
+                it("should silently skip virtual files that match only universal patterns", async () => {
+                    eslint = new ESLint({
+                        overrideConfigFile: true,
+                        overrideConfig: [
+                            pluginConfig,
+                            {
+                                files: ["**/*"],
+                                rules: {
+                                    "no-undef": 2
+                                }
+                            }
+                        ]
+                    });
+                    const [result] = await eslint.lintText(text, { filePath: "foo.md" });
+
+                    assert.strictEqual(result.messages.length, 3);
+                    assert.strictEqual(result.messages[0].ruleId, "no-undef");
+                    assert.match(result.messages[0].message, /foo_js/u);
+                    assert.strictEqual(result.messages[0].line, 2);
+                    assert.strictEqual(result.messages[1].ruleId, "no-undef");
+                    assert.match(result.messages[1].message, /foo_cjs/u);
+                    assert.strictEqual(result.messages[1].line, 10);
+                    assert.strictEqual(result.messages[2].ruleId, "no-undef");
+                    assert.match(result.messages[2].message, /foo_mjs/u);
+                    assert.strictEqual(result.messages[2].line, 14);
+                });
+
+                it("should silently skip virtual files that are ignored by global ignores", async () => {
+                    eslint = new ESLint({
+                        overrideConfigFile: true,
+                        overrideConfig: [
+                            pluginConfig,
+                            {
+                                rules: {
+                                    "no-undef": 2
+                                }
+                            },
+                            {
+                                ignores: ["**/*.cjs"]
+                            }
+                        ]
+                    });
+                    const [result] = await eslint.lintText(text, { filePath: "foo.md" });
+
+                    assert.strictEqual(result.messages.length, 2);
+                    assert.strictEqual(result.messages[0].ruleId, "no-undef");
+                    assert.match(result.messages[0].message, /foo_js/u);
+                    assert.strictEqual(result.messages[0].line, 2);
+                    assert.strictEqual(result.messages[1].ruleId, "no-undef");
+                    assert.match(result.messages[1].message, /foo_mjs/u);
+                    assert.strictEqual(result.messages[1].line, 14);
+                });
+
+                // https://github.com/eslint/eslint/issues/15949
+                it("should silently skip virtual files that are ignored by global ignores even if they match non-universal patterns", async () => {
+                    eslint = new ESLint({
+                        overrideConfigFile: true,
+                        overrideConfig: [
+                            pluginConfig,
+                            {
+                                rules: {
+                                    "no-undef": 2
+                                }
+                            },
+                            {
+                                files: ["**/*.ts"]
+                            },
+                            {
+                                ignores: ["**/*.md/*.ts"]
+                            }
+                        ]
+                    });
+                    const [result] = await eslint.lintText(text, { filePath: "foo.md" });
+
+                    assert.strictEqual(result.messages.length, 3);
+                    assert.strictEqual(result.messages[0].ruleId, "no-undef");
+                    assert.match(result.messages[0].message, /foo_js/u);
+                    assert.strictEqual(result.messages[0].line, 2);
+                    assert.strictEqual(result.messages[1].ruleId, "no-undef");
+                    assert.match(result.messages[1].message, /foo_cjs/u);
+                    assert.strictEqual(result.messages[1].line, 10);
+                    assert.strictEqual(result.messages[2].ruleId, "no-undef");
+                    assert.match(result.messages[2].message, /foo_mjs/u);
+                    assert.strictEqual(result.messages[2].line, 14);
                 });
             });
         });

